### ISSUE TITLE
unused_mut warning in macro invocation, although mut is needed in other invocation

### ISSUE THREAD

**Issue by hellow554** [rust-lang/rust#96547]
### Code

I tried this code ([Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=e7ac1f9f412334732cf8e86725842b9d)):

```rust
macro_rules! check_ok_fun {
    ($val: expr, $expected: expr) => {{
        let v = $val;
        let mut e = $expected;

        assert!(e(v));
    }};
}

fn main() {
    let mut a = None;
    check_ok_fun!(3, |_| true);
    check_ok_fun!(3, |_| {
        a = Some(3);
        true
    });
}
```

The current output is:

```
warning: variable does not need to be mutable
  [--> src/main.rs:4:13
](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=e7ac1f9f412334732cf8e86725842b9d#)   |
4  |         let mut e = $expected;
   |             ----^
   |             |
   |             help: remove this `mut`
...
12 |     check_ok_fun!(3, |_| true);
   |     -------------------------- in this macro invocation
   |
   = note: `#[warn(unused_mut)]` on by default
   = note: this warning originates in the macro `check_ok_fun` (in Nightly builds, run with -Z macro-backtrace for more info)
```

Ideally the output should look like:

No warning

---

The problem is, that when I remove the `mut` the second `check_ok_fun!` won't work anymore.

**Comment by wwylele** [comment]
Diagnostics is done after macro expansion, on per-invocation basis. This is why macro is a legit and heavy user of `#[allow(...)]` annotation to silent harmless diagnostic warnings.
